src/nvim/plines.c

```diff
@@ -146,7 +146,7 @@ CharSize charsize_regular(CharsizeArg *csarg, char *const cur, colnr_T const vco
   } else if (cur_char < 0) {
     size = kInvalidByteCells;
   } else {
-    size = char2cells(cur_char);
+    size = ptr2cells(cur);
```

Comment:
This kind of makes it pointless to pass in `cur_char`, as `utf_ptr2cells()` already handles illegal bytes, and the `cur_char > 0x80` check can be replaced with `MB_BYTE2LEN(*cur) > 1`.

Or the logic in `utf_ptr2cells()` can be replicated here without the first `utf_ptr2char()` to avoid decoding first char twice, but then that will also require passing in `ci.chr.len`, so not sure if that's worth it.

Reply:
ye, multiple decoding also happens in other places like the main win_line() loop. we probably want a specialized version of CharInfo which also includes the ptr2cells() width calculated at the same time as the byte length. Although I am thinking of that as a follow-up perf PR while only focusing on correctness (and no larger regressions) in this PR.